src/_pytest/tmpdir.py

```diff
@@ -47,6 +48,9 @@ def from_config(cls, config) -> "TempPathFactory":
 
     def mktemp(self, basename: str, numbered: bool = True) -> Path:
         """makes a temporary directory managed by the factory"""
+
+        if os.path.isabs(basename):
+            fail("Absolute and non-normalized path is disallowed: {}".format(basename))
```

Comment:
we need to compare basename with its normpath to be sure of that

i beleive a `ValueError` is better here, i apologize for using the word `fail` as it is ambiguous in the pytest context and i meant it in the general sense instead of `pytest.fail`

as for the message, i would propose `{} is not a normalized and relative path`

Reply:
> as for the message, i would propose {} is not a normalized and relative path

Why "normalized", wouldn't `{} is not a relative path` suffice? Or you want to prevent `"../../foo"` from being passed?

Reply:
Exactly, malicious testsuites shouldn't be able to misuse good looking api to hide intent

Reply:
I see, thanks.

But `normpath` is not enough:

```
>>> import os.path
>>> os.path.normpath('..')
'..'
>>> os.path.normpath('../foo')
'..\\foo'
```

We probably want to check if the path also starts with `..`:

```python
basename = os.path.normpath(basename)
if os.path.isabs(basename) or basename.startswith(".."):
    ...
```

If that is correct, we probably should isolate this logic into a function so we can document it and unittest it properly.

Reply:
maybe not just starts with "..", also below case
```python
>>> os.path.normpath("../../hello")
'..\\..\\hello'
```

Reply:
can we implement like below

```python

...     if os.path.isabs(basename) or ".." in basename:
...             raise ValueError("{} is not a normalized and relative path".format(basename))
```

Reply:
> maybe not just starts with "..", also below case

But your example starts with `..`. 😉 

It checking the normalized path starting with `..` should be enough, do you have examples where that's not true?

Reply:
```python
testdata = [
    "hello",
    "/hello",
    "../hello",
    "../hello/../hello",
    "\\hello",
    "..\\hello",
    "hello\\..",
    "hello\\..\\hello",
]
```

Reply:
Sorry I don't follow... are you calling `os.path.normpath` before checking if it starts with `..`?

Reply:
shouldn't we let the user to normalize the path? and let pytest raise exception if it is absolute or non-normalized?

if calling `os.path .normpath` before checking, then results below
 **run on windows**

```python
FAILED experiment.py::test_disallow[/hello1] - ValueError: disallow abs and non-normalized paths
FAILED experiment.py::test_disallow[../hello2] - ValueError: disallow abs and non-normalized paths
FAILED experiment.py::test_disallow[\\hello4] - ValueError: disallow abs and non-normalized paths
FAILED experiment.py::test_disallow[..\\hello5] - ValueError: disallow abs and non-normalized paths
FAILED experiment.py::test_disallow[hello6\\..] - FileExistsError: [WinError 183] Cannot create a file when that file already exists: 'C:\\Us..
```

**run on linux**
```python
FAILED experiment.py::test_disallow[/hello] - ValueError: disallow abs and non-normalized paths
FAILED experiment.py::test_disallow[../hello] - ValueError: disallow abs and non-normalized paths
FAILED experiment.py::test_disallow[..\\hello] - ValueError: disallow abs and non-normalized paths
```

Reply:
Given how tricky this turn out, we should check for absolute paths and for escapist normalitzd paths, I always forget how painfully fragile path handling is

Reply:
Weird, I don't get the same results at all:

```python
>>> os.path.normpath('/hello')
'\\hello'
```

Can you post `test_disallow.py` please?

Reply:
if use "startswith", then it allows `"hello\\..",` `"hello\\..\\hello",` to pass on linux, if we always check if any ".." in str, then we can disallow above two cases on linux

should we do below? please comment
```python
    basename = os.path.normpath(basename)
    if os.path.isabs(basename) or ".." in basename or basename.startswith("\"):
        raise ValueError("disallow abs and non-normalized paths")
```

Reply:
here is my test codes

```python
import os.path
import pytest
import py


testdata = [
    "hello",
    "/hello1",
    "../hello2",
    "hello3/../hello3/",
    "\\hello4\\",
    "..\\hello5\\",
    "hello6\\..",
    "hello7\\..\\hello7",
]


@pytest.mark.parametrize("basename", testdata)
def test_disallow(basename, tmpdir_factory):
    basename = os.path.normpath(basename)
    print("DEBUG: {}".format(basename))
    if os.path.isabs(basename) or basename.startswith(".."):
        raise ValueError("disallow abs and non-normalized paths")
    tmpdir_factory.mktemp(basename, numbered=False)


@pytest.mark.parametrize("basename", testdata)
def test_disallow_2(basename, tmpdir_factory):
    basename = os.path.normpath(basename)
    print("DEBUG: {}".format(basename))
    if os.path.isabs(basename) or ".." in basename:
        raise ValueError("disallow abs and non-normalized paths")
    tmpdir_factory.mktemp(basename, numbered=False)


```

Reply:
Thanks! Updated the example a bit:

```python
import os.path
from contextlib import nullcontext

import pytest

testdata = [
    ("hello", True),
    ("/hello1", False),
    ("../hello2", False),
    ("hello3/../hello3/", True),
    ("\\hello4\\", False),
    ("..\\hello5\\", False),
    ("hello6\\..", False),
    ("hello7\\..\\hello7", True),
]


@pytest.mark.parametrize("basename, ok", testdata)
def test_disallow(basename, tmpdir_factory, ok):

    def check(basename):
        basename = os.path.normpath(basename)
        if os.path.isabs(basename) or basename.startswith(".."):
            raise ValueError("disallow abs and non-normalized paths")
        
    ctx = nullcontext() if ok else pytest.raises(ValueError)
    with ctx:
        check(basename)
```

And only `"hello6\\.."` fails.

But this seems pretty brittle, perhaps we should resolve the full path and see that it is contained within `tempdir`?

Reply:
can you also test on linux?

Reply:
Not easily right now, but I'm pretty sure we will get the same results (althought we need to use `os.sep` in place of `\\`).

But resolving the path name and checking if it is a subdir should work (untested):

```
if Path(basename).resolve().parent != self.getbasetemp():
    raise ValueError(...)
```

Reply:
on windows, `hello6\\..` will create "." directory and will fail
I tested your solution on linux, 

```
experiment.py::test_disallow[hello-True] PASSED                                                                  [ 12%]
experiment.py::test_disallow[/hello1-False] PASSED                                                               [ 25%]
experiment.py::test_disallow[../hello2-False] PASSED                                                             [ 37%]
experiment.py::test_disallow[hello3/../hello3/-True] PASSED                                                      [ 50%]
experiment.py::test_disallow[\\hello4\\-False] FAILED                                                            [ 62%]
experiment.py::test_disallow[..\\hello5\\-False] PASSED                                                          [ 75%]
experiment.py::test_disallow[hello6\\..-False] FAILED                                                            [ 87%]
experiment.py::test_disallow[hello7\\..\\hello7-True] PASSED                                                     [100%]

```

Reply:
@RonnyPfannschmidt @nicoddemus , my question is 
"Shouldn't we let the user to normalize the path instead we do it inside pytest?"
Then, we only need to check if it is absolute  or non-normalized path.

Reply:
for security/sanity/ease of use, we should normalize, and fail if it tries to escape/is absolute

Reply:
> on windows, hello6\\.. will create "." directory and will fail

Hmm how? `Path("hello6\\..").resolve().parent` will return the parent of basetemp, so it should be different than `basetemp()`.

Could you also post the test script you are using besides the output, so we can adjust accordingly?

Reply:
```python

import os.path
from contextlib import nullcontext

import pytest
from pathlib import Path

testdata = [
    ("hello", True),
    ("/hello1", False),
    ("../hello2", False),
    ("hello3/../hello3/", True),
    ("\\hello4\\", False),
    ("..\\hello5\\", False),
    ("hello6\\..", False),
    ("hello7\\..\\hello7", True),
]


@pytest.mark.parametrize("basename, ok", testdata)
def test_disallow(basename, testdir, tmpdir_factory, ok):
    def check(basename):
        basename = os.path.normpath(basename)
        print("DEBUG parent: ", Path(basename).resolve().parent, end=";;  ")
        print("DEBUG basetemp: ", tmpdir_factory.getbasetemp(), end=" ")

        if (
            os.path.isabs(basename)
            or basename.startswith("..")
            or Path(basename).resolve().parent != tmpdir_factory.getbasetemp()
        ):
            raise ValueError("disallow abs and non-normalized paths")

    ctx = nullcontext() if ok else pytest.raises(ValueError)
    with ctx:
        check(basename)

```

Reply:
Thanks! The problem was that `basename` was not being joined to `getbasetemp()`. Here's the full working code:

```python
import os.path
from contextlib import nullcontext

import pytest
from pathlib import Path

testdata = [
    ("hello", True),
    ("/hello1", False),
    ("../hello2", False),
    ("hello3/../hello3/", True),
    ("\\hello4\\", False),
    ("..\\hello5\\", False),
    ("hello6\\..", False),
    ("hello7\\..\\hello7", True),
]

@pytest.mark.parametrize("basename, ok", testdata)
def test_disallow(basename, testdir, tmp_path_factory, ok):
    def check(basename):
        full_path = tmp_path_factory.getbasetemp() / basename
        if full_path.resolve().parent != tmp_path_factory.getbasetemp():
            raise ValueError("disallow abs and non-normalized paths")

    ctx = nullcontext() if ok else pytest.raises(ValueError)
    with ctx:
        check(basename)


```

Reply:
all test pass on windows, but not on linux. 

**result on linux**, the test "\\hello4\\" and "hello6\\.." will fail

experiment.py::test_disallow[hello-True] PASSED
experiment.py::test_disallow[/hello1-False] PASSED
experiment.py::test_disallow[../hello2-False] PASSED
experiment.py::test_disallow[hello3/../hello3/-True] PASSED
**experiment.py::test_disallow[\\hello4\\-False] FAILED**
experiment.py::test_disallow[..\\hello5\\-False] PASSED
**experiment.py::test_disallow[hello6\\..-False] FAILED**
experiment.py::test_disallow[hello7\\..\\hello7-True] PASSED


**below are the codes to test**

```python
import os.path
from contextlib import nullcontext

import pytest
from pathlib import Path

testdata = [
    ("hello", True),
    ("/hello1", False),
    ("../hello2", False),
    ("hello3/../hello3/", True),
    ("\\hello4\\", False),
    ("..\\hello5\\", False),
    ("hello6\\..", False),
    ("hello7\\..\\hello7", True),
]


@pytest.mark.parametrize("basename, ok", testdata)
def test_disallow(basename, testdir, tmp_path_factory, ok):
    def check(basename):
        basename = os.path.normpath(basename)
        full_path = tmp_path_factory.getbasetemp() / basename

        print("DEBUG parent: ", full_path.resolve().parent, end=";;  ")
        print("DEBUG basetemp: ", tmp_path_factory.getbasetemp(), end=" ")
        if (
            os.path.isabs(basename)
            or basename.startswith("..")
            or full_path.resolve().parent != tmp_path_factory.getbasetemp()
        ):
            raise ValueError("disallow abs and non-normalized paths")

    ctx = nullcontext() if ok else pytest.raises(ValueError)
    with ctx:
        check(basename)

```

Reply:
it is tricky to make it completely platform agnostic for all cases.
I think it is good to use your original solution.
```python
basename = os.path.normpath(basename)
if os.path.isabs(basename) or basename.startswith(".."):
       raise ValueError("disallow abs and non-normalized paths")
```

Reply:
Actually the failures on Linux are fine, `\\` is not supposed to work on it anyway. 

I suggest to change all the cases with `\\` to `/`, as `/` works on Windows as well.

Reply:
OK, sounds reasonable, I changed the test data a bit. Is below implementation OK? Please comment if I can go ahead with new PR

```python
testdata = [
    ("hello", True),
    ("/hello1", False),
    ("./hello2", True),
    ("../hello3", False),
    ("../../hello4", False),
    ("hello5/..", False),
    ("hello6/../hello6", True),
]


@pytest.mark.parametrize("basename, ok", testdata)
def test_disallow(basename, testdir, tmp_path_factory, ok):
    def check(basename):
        basename = os.path.normpath(basename)
        full_path = tmp_path_factory.getbasetemp() / basename

        # print("DEBUG parent: ", full_path.resolve().parent, end=";;  ")
        # print("DEBUG basetemp: ", tmp_path_factory.getbasetemp(), end=" ")
        if (
            os.path.isabs(basename)
            or basename.startswith("..")
            or full_path.resolve().parent != tmp_path_factory.getbasetemp()
        ):
            raise ValueError("disallow abs and non-normalized paths")

    ctx = nullcontext() if ok else pytest.raises(ValueError)
    with ctx:
        check(basename)
```

Reply:
LGTM, I would just also add `("hello6/../hello6/..", False),` to the list of cases.

Reply:
what is LGTM:)

Reply:
It's an acronym for "looks good to me".

Reply:
Yep, sorry about that. Btw a site I use to check for acronyms is acronymfinder.com. 😉